### DISCUSSION TITLE
Deprecated: JPA Map Storage: How To and Q&A

### DISCUSSION THREAD

**Discussion by ahus1** [keycloak/keycloak#9596] [Archive]
This post summarizes what the JPA Map Storage provider can do today, how it can be tested and configured, and things to watch out. Comments provided to in this discussion will be used to update the original post. The post aims to reflect the current state of development in Keycloak's main branch. 

For a discussion what this store could do in the future, consider creating a new discussion as this post focuses on how to test and use what is already on the main branch.

## Audience, Goals, Status

**Audience:** (will change while development progresses)
* Developers who want to explore and contribute to this store
* Future to-be-users that want to give this store a test run in a local development environment

**Goal of the Map Storage Provider Family:** Plugable storage back ends for the different first-class objects that will eventually support zero downtime upgrades, see https://github.com/keycloak/keycloak-community/blob/main/design/keycloak.x/storage-architecture.md

**Goal of the JPA Map Storage Provider:** Map storage provider backed by an SQL database. First-class objects stored as JSON to support lazy schema migration with no downtime, with additional columns/tables to support efficient retrieval and search, plus database consistency via constraints within an entity (but not between entities).

**Status:** early preview, see announcement here: https://www.keycloak.org/2022/07/storage-map.html

**Supported Objects:** All stores as listed below. Can be activated independently for development, and all stores active for the Quarkus distribution.

**Supported Databases:** PostgreSQL, CockroachDB (other databases will eventually follow)

**Supported Isolation Levels:** 
* With PostgreSQL, transactions run at the isolation level READ_COMMITTED. Request that modify objects concurrently might fail due to optimistic locking exceptions when running in READ_COMMITTED isolation level. The GitHub Action test suite passes for PostgreSQL since the KC19 release.
* With CockroachDB, transactions run at the isolation level SERIALIZABLE. The GitHub Action test suite passes for CockroachDB for the upcoming KC21 release.

**Upgrades from previous versions:** While this is in development, no upgrades are supported. When trying out a newer version, always start with an empty realm setup / empty database (all tables dropped).

## Setup of a test environment (minimal)

* Setup of a PostgreSQL database using either docker or podman: `podman run --name postgres -p 5432:5432 -e POSTGRES_PASSWORD=pass -e POSTGRES_USER=keycloak -e POSTGRES_DB=keycloak -d postgres:13.2`

Use host localhost, port 5432, database `keycloak`, user `keycloak` and password `pass` to connect to the database later to debug or to inspect the database.

* Clear data from PostgreSQL database (needed when there have been changes to the changesets and checksums are out of sync)
```
drop schema public cascade;
create schema public;
``` 

* If a mixture of JPA Map store and Concurrent Hash Map is used: Clear data from the concurrent hash map storage (needs to be cleared if the database is being cleared)
```
rm -rf target/map
```

* Build Keycloak (needed when there have been changes to Keycloak source code)
```
mvn clean install -DskipTests
```

* Run `org.keycloak.testsuite.KeycloakServer` with the following parameters; the given properties will replace values in  https://github.com/keycloak/keycloak/blob/main/testsuite/utils/src/main/resources/META-INF/keycloak-server.json
```
mvn -pl testsuite/utils \
-Pkeycloak-server,map-storage \
-Dkeycloak.authSession.map.storage.provider=jpa \
-Dkeycloak.client.map.storage.provider=jpa \
-Dkeycloak.clientScope.map.storage.provider=jpa \
-Dkeycloak.user.map.storage.provider=jpa \
-Dkeycloak.role.map.storage.provider=jpa \
-Dkeycloak.realm.map.storage.provider=jpa \
-Dkeycloak.group.map.storage.provider=jpa \
-Dkeycloak.loginFailure.map.storage.provider=jpa \
-Dkeycloak.authorization.map.storage.provider=jpa \
-Dkeycloak.userSession.map.storage.provider=jpa \
-Dkeycloak.singleUseObject.map.storage.provider=jpa \
-Dkeycloak.actionToken.map.storage.provider=jpa \
-Dkeycloak.adminEventsStore.map.storage.provider=jpa \
-Dkeycloak.userEventsStore.map.storage.provider=jpa \
-Dkeycloak.map.storage.connectionsJpa.url=jdbc:postgresql://localhost:5432/keycloak \
-Dkeycloak.map.storage.connectionsJpa.user=keycloak \
-Dkeycloak.map.storage.connectionsJpa.password=pass \
exec:java
``` 

* Verify setup after start: Connect to the database. Keycloak tables (`kc_client`, `kc_client_attribute`, `kc_client_scope`, etc.) and `databasechangelog` should exist. All tables should contain data from the initially set up master realm.

**Alternative setup with Quarkus**

After building Keycloak, run 

```
mvn -f quarkus/server/pom.xml compile quarkus:dev -Dquarkus.args="start-dev --storage=jpa --db-password=pass --db-username=keycloak --db-url=jdbc:postgresql://localhost:5432/keycloak" -Dkc.config.args="start-dev;;--storage=jpa;;"
```

Or use the [Keycloak nightly distribution ](https://github.com/keycloak/keycloak/releases/tag/nightly) or unpack the locally build file `quarkus/dist/target/keycloak-999-SNAPSHOT.zip` and run 

```
bin/kc.sh start-dev --storage=jpa --db-password=pass --db-username=keycloak --db-url=jdbc:postgresql://localhost:5432/keycloak
```

**Alternative setup with CockroachDB**

* Setup a local CockroachDB single-node instance using docker
    * Start single node: `docker run -d --name=cockroach -p 26257:26257 -e COCKROACH_USER=keycloak -e COCKROACH_DATABASE=keycloak  -p 8280:8080 cockroachdb/cockroach:latest-v22.1 start-single-node --insecure`

* Run `org.keycloak.testsuite.KeycloakServer` as described above. The only changed needed is the `JDBC` URL, that must be adjusted to use port `26257`. So the property must read

```
-Dkeycloak.map.storage.connectionsJpa.url=jdbc:postgresql://localhost:26257/keycloak
```
CockroachDB uses the PostgreSQL driver for JDBC connections and this leads to minimal adjustments to make `KeycloakServer` run with CockroachDB instead of PostgreSQL.

* Verify setup after start: Connect to the database (run Cockroach's `sql` tool and then run `\c keycloak;` to connect to the `keycloak` database). Then type `\d` to see that all tables were created successfully.

* Clear data from CockroachDB  (needed for subsequent restarts):
    * Connect to the `defaultdb` database if you are currently connected to the `keycloak` database: `\c defaultdb`
    * Drop the `keycloak` database: `DROP DATABASE keycloak CASCADE;`
    * Re-create the `keycloak` database: `CREATE DATABASE keycloak`;
    * Grant privileges on the new DB to user `keycloak`: `GRANT ALL ON DATABASE keycloak TO keycloak;`

## Known issues and limitations

[Open Issues](https://github.com/orgs/keycloak/projects/9/views/1?filterQuery=is%3Aissue++-status%3A%22%E2%9C%85+Done%22+new-store-scope%3A%22%F0%9F%97%84+JPA+%28physical+layer%29%22)

**Comment by vramik** [comment]
For clearing data from database it could also be used following (assuming the database uses schema `public`)

```
drop schema public cascade;
create schema public;
```

**Reply by ahus1** [reply]
Thanks - updated the description. Also added that the roles are now supported.

**Comment by ahus1** [comment]
Updated so reflect that client_scopes are now supported. Added a paragraph about isolation levels.

**Comment by ahus1** [comment]
Updated to reflected added support for groups. Table names for Keycloak now have the prefix "kc_".

**Comment by sguilhen** [comment]
We'll be adding a section on how to alternatively use CockroachDB.

**Reply by ahus1** [reply]
Thank - I added the section to he description above!

**Reply by ahus1** [reply]
@sguilhen / @vramik - when I was working with CockroachDB last week I found that (at least in 21.1) there is the option "start-single-node". For me it simplified the setup to not have a volume and also to skip the initialization. 

I updated the docs above, I hope they will allow more people to try out CockroachDB.

**Comment by vramik** [comment]
@ahus1 FYI https://github.com/keycloak/keycloak/issues/12098

**Reply by ahus1** [reply]
Thanks for letting me know, will update the docs once it has been merged.

**Reply by ahus1** [reply]
@kami619 - I've updated the CLI to configure the storage; instead of "jpa-map-storage" it is now "jpa" as changed in the PR mentioned above.

**Comment by ahus1** [comment]
As #9669 has been completed, JPA map storage is now supporting authorizations as well, and the command line above have been updated. Thanks @vramik for the implementation!

**Comment by sguilhen** [comment]
@ahus1 Just realized that the user login failures store has been available for a while now and is missing here. Relevant properties:

```
-Dkeycloak.loginFailure.provider=map
-Dkeycloak.loginFailure.map.storage.provider=jpa
```

**Reply by ahus1** [reply]
Thank you for that hint - updated the run configuration! I wonder: should I instead of listing all the jpa stores separately, create a new profile like `map-storage-jpa` that sets all the currently supported providers? I would look less impressive, but would be simpler to maintain.

**Reply by sguilhen** [reply]
> Thank you for that hint - updated the run configuration! I wonder: should I instead of listing all the jpa stores separately, create a new profile like `map-storage-jpa` that sets all the currently supported providers? I would look less impressive, but would be simpler to maintain.

This gets a +1 from me. We already have that for the testsuite (in `integration-arquillian/tests/base/pom.xml`)

**Comment by ahus1** [comment]
Updated the description on how to run JPA Map storage with Quarkus

**Comment by vramik** [comment]
**Known limitations**

- Supported databases: PostgresSQL, CockroachDB
- CockroachDB - test failures: https://github.com/keycloak/keycloak/issues/11618


@ahus1 the note regarding isolation level should be probably removed, when I was testing it with SERIALIZABLE and READ_COMMITTED I've seen failures (due to the fact it was not possible to serialize transactions)

**Supported Objects** also seems outdated.

**Reply by ahus1** [reply]
Hi @vramik - thank you, I updated the description in the areas you mentioned. Looking at the code, I see that the usersession PR removed the configurability for the isolation level, and it now defaults to the respective default of the database (PostgreSQL = READ COMMITTED, CockroachDB = SERIALIZABLE).

**Reply by ahus1** [reply]
Update the document again as the test suite for CockroachDB now passes (thank you, @sguilhen!)

**Comment by DGuhr** [comment]
**Question:** How can I see all the available CLI options for the new store directly from the cli?
**Answer:** invoke `kc.sh start/start-dev/build --help-all` ;) just read through the blog post and we forgot to mention that they don't show up using only `--help`

**Comment by ThoreKr** [comment]
After todays update, will there be an additional followup to this blog post https://www.keycloak.org/2022/02/dbs.html lining out how this will affect future database compatibility?

In particular I was looking forward to cockroach compatibility. Are there already plans which databases will be supported in the future?

**Reply by ahus1** [reply]
For now, we will keep the compatibility with the existing databases of the current store. At some point we were hoping people would move more towards PostgreSQL databases, but we still see a great demand for other databases in enterprises. 

We might revisit adding CockroachDB support for the current store eventually, but only if it adds a significant value compared to the effort we need to put into supporting it. IMHO to make use of the full capabilities of CockroachDB for multi-region setups, one would both need Cockroach enterprise features, additional columns in the database to support for example sharing of the data in different regions. So the effort we need to put into understanding this and supporting this is significant, and we will not take this lightly. Maybe there is a discussion/issue of this topic already. If so, please link it here and continue the discussion over there, or create a new discussion.

**Comment by ahus1** [comment]
As announced to the community in [this blog post](https://www.keycloak.org/2023/10/map-store-removal), we are no longer developing the Map Store, and it will be removed from Keycloak's codebase. Follow #24111 for the removal steps.
